Frameworks/UIKit/_UIGestureCoordinator.mm

```diff
@@ -0,0 +1,457 @@
+        }
+    }
+}
+
+// main entrance to send touch to gesture pipleline for processing
+- (bool)processGesturesForTouch:(UITouch*)touch event:(UIEvent*)event touchEventName:(SEL)eventName {
+    if (DEBUG_GESTURES) {
+        TraceVerbose(TAG, L"processGesturesForTouch.");
+    }
+
+    [self _buildTrackingGesturesListForView:touch.view];
+
+    // Send touch to regualr gesture for state transiton
+    BOOL dManipGesturesCancelled = NO;
+    for (UIGestureRecognizer* curgesture in _regularRecognizers) {
+        if ([curgesture state] != UIGestureRecognizerStateCancelled) {
+            if (DEBUG_GESTURES) {
+                TraceVerbose(TAG, L"Checking gesture %hs.", object_getClassName(curgesture));
+            }
+
+            id delegate = [curgesture delegate];
+            BOOL send = TRUE;
+            if (touch.phase == UITouchPhaseBegan && [delegate respondsToSelector:@selector(gestureRecognizer:shouldReceiveTouch:)]) {
+                send = [delegate gestureRecognizer:curgesture shouldReceiveTouch:touch];
+            }
+
+            if (send) {
+                [curgesture performSelector:eventName withObject:[NSMutableSet setWithObject:touch] withObject:event];
+                // Verify if gesture state change to Began or recognized after transition, which means a gesture just become active
+                if ((curgesture.state == UIGestureRecognizerStateBegan) || (curgesture.state == UIGestureRecognizerStateRecognized)) {
+                    // Remove any gesture in possible state which can not fire concurrently with newly active gesture
+                    [self _removeGesturesByActiveGesture:curgesture];
+
+                    if (!dManipGesturesCancelled) {
+                        // Cancel all DimanipGesture as well if have not done so
+                        [self _cancelAllDManipGestures];
+                        dManipGesturesCancelled = YES;
+                    }
+
+                    if (DEBUG_GESTURES) {
+                        TraceVerbose(TAG,
+                                     L"gesture %hs is in state %d, cancel DManipGesture.",
+                                     object_getClassName(curgesture),
+                                     curgesture.state);
+                    }
+                }
+            }
+        }
+    }
+
+    if (!dManipGesturesCancelled) {
+        // if DimanipGesture has not been cancelled, send touch to it for state transition
+        for (_UIDMPanGestureRecognizer* dManipGesture in _dManipRecognizers) {
+            if ([dManipGesture state] != UIGestureRecognizerStateCancelled) {
+                id delegate = [dManipGesture delegate];
+
+                BOOL send = TRUE;
+                if (touch.phase == UITouchPhaseBegan && [delegate respondsToSelector:@selector(gestureRecognizer:shouldReceiveTouch:)]) {
+                    send = [delegate gestureRecognizer:dManipGesture shouldReceiveTouch:touch];
+                }
```

Comment:
code refactored

---
In reply to: [100627188](https://github.com/Microsoft/WinObjC/pull/1909#discussion_r100627188) [](ancestors = 100627188)